README.md
---

## uxcore-date-form-field [![Dependency Status](http://img.shields.io/david/uxcore/uxcore-date-form-field.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-date-form-field) [![devDependency Status](http://img.shields.io/david/dev/uxcore/uxcore-date-form-field.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-date-form-field#info=devDependencies) 

## TL;DR

uxcore-date-form-field ui component for react

#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-date-form-field
$ cd uxcore-date-form-field
$ npm install
$ npm run start
```

## Usage

```
<Form
    jsxmode={this.state.mode}
    ref={(c) => { this.form = c; }}
    jsxvalues={{
        testDate: '',
        casDate: ['2016-01-02', '2016-02-03'],
    }}
>
    <DateFormField
        showTime={false}
        format="yyyy-MM-dd"
        jsxname="date"
        jsxlabel="日期"
        locale="zh-cn"
        className="testDate"
    />
    <DateFormField
        jsxtype="cascade"
        autoMatchWidth
        useFormat
        format="yyyy-MM-dd"
        jsxfrom="2016-11-24"
        jsxname="format"
        jsxlabel="日期区间"
        locale="zh-cn"
        required
        requireType="end"
        jsxrules={[
            { validator: Validators.isNotEmpty, errMsg: '不能为空' },
        ]}
    />
</Form>
```

更多请参阅 Form 文档。

## Props

支持 FormField 公共的 Props 之外还支持以下 Props：

| 配置项 | 类型 | 必填 | 默认值 | 功能/备注 |
|---|---|---|---|---|
|jsxtype|string|optional|single|single/cascade 单独、级联|
|jsxfrom|string|optional|-|开始日期|
|jsxto|string|optional|-|结束日期|
|panel|string|optional|"day"|何种面板,枚举值为"month","year"和"day"|
|autoMatchWidth|boolean|optional|false|从 1.8.11 版本支持，在级联状态下输入框自动匹配宽度|
|requireType|string|optional|-|必填校验方式，有4个可选值，见 [requireType](#requireType) 说明|

除此之外，支持除 onSelect，[uxcore-calendar](https://www.npmjs.com/package/uxcore-calendar) 的所有 Props。

### requireType

用于定义在级联模式下的必填校验规则，默认不传的情况的校验逻辑是 any，即任意填写一项即可。

- start 区间开始时间必填
- end 区间结束时间必填
- both 区间开始时间和结束时间均必填
- any 区间开始时间和结束时间任意填写一项

HISTORY.md
# history
## 0.9.13
* `FIXED` fix a bug in disabledDate when showTime is true

## 0.9.12
* `FIXED` add new prop fixTimezoneOffset for view mode

## 0.9.11
* `FIXED` add defaultValue support in cascade mode

## 0.9.10
* `CHANGE` when select a quickSelectRanges, close the tooltip

## 0.9.8
* `NEW` add new prop quickSelectRanges

## 0.9.3
* `NEW` add metadata support

## 0.9.2

* `FIXED` resize should be called if panel is changed

## 0.9.1

* `FIXED` autoMatchWidth fails to work if jsxshow is changed from false to true

## 0.9.0

* `CHANGED` upgrade `uxcore-form-field` to `^0.3.0`
* `CHANGED` upgrade `uxcore-calendar` to `^0.10.0`
* `CHANGED` adapt React 16

## 0.8.13

* `CHANGED` spliter style change

## 0.8.11

* `FIXED` more reliable autoMatchWidth if style is loaded later than component

## 0.8.10

* `CHANGED` new autoMatchWidth, more reliable

## 0.8.9

* `FIXED` cascadeBox width change cannot trigger resize when update

## 0.8.8

* `FIXED` autoSize logic go wrong if cascade wrapper's width is not equal to fieldCore's width
* `CHANGED` `inputBoxMaxWidth`, `mode` & `verticalAlign` change will call the resize

## 0.8.7

* `CHANGED` add a wrapper for cascade type

## 0.8.6

* `CHANGED` fit React@15

## 0.8.5

* `CHANGED` support size config

## 0.8.4

* `FIXED` pass `className` to `Calendar`

## 0.8.3

* `FIXED` resize bug in view mode

## 0.8.0

* `CHANGED` update `uxcore-calendar` to ~0.9.0

## 0.7.3

* `FIXED` view mode will show 1970-01-01 if value is null

## 0.7.2

* `CHANGED` use different default format in `year` & `month` panel

## 0.7.1

* `CHANGED` show key if label is not found in view mode

## 0.7.0

* `CHANGED` update `uxcore-calendar` to ~0.8.0

## 0.6.1

* `FIXED` `processTime` timezone bug (also fix in ver. 0.4.x ~ 0.5.x)

## 0.6.0

* `CHANGED` update `uxcore-calendar` to ~0.7.0

## 0.5.2

* `FIXED` use parseInt to get core width, fix the width bug in IE 11.

## 0.5.1

`CHANGED` optimize resize performance

## 0.5.0

`NEW` add new prop `autoMatchWidth`
`CHANGED` update formField to ~0.2.0

## 0.4.5

`FIXED` spread attributes with undefined key will override the previous attribute [#4](https://github.com/uxcore/uxcore-date-form-field/issues/4)

## 0.4.4

`CHANGED` use `processTime` in `handleCascadeChange` & `handleChange` method.

## 0.4.3

`CHANGED` cascade mode support `disabledDate`

## 0.4.2

`CHANGED` `format` support array

## 0.4.1

`CHANGED` js style standardization
`CHANGED` `disabled` `placeholder` support array

## 0.4.0

`CHANGED` update uxcore-calendar to ~0.6.0

## 0.3.2

`FIX` date before 1970.01.01 cannot be selected.

## 0.3.1

`NEW` support new prop `useFormat`, DateFormField will use formatted value to communicate.

## 0.3.0

`CHANGED` update dependency `uxcore-calendar` to ~0.5.0

## 0.2.3

`FIXED` fix bug when using showDateInput

## 0.2.2

`NEW` add new prop `panel` to support MonthPanel and YearPanel

## 0.2.1

`FIX` fix server render

## 0.2.0

`CHANGED` update uxcore-calendar to ~0.4.0

## 0.1.3

`CHANGED` fix view mode bug (showing timestamp)

## 0.1.2

`CHANGED` keep deepcopy version same with `uxcore-form`


tests/DateFormField.spec.js
import expect from 'expect.js';
import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import Form from 'uxcore-form/build/Form';
import Constants from 'uxcore-const';
import $ from 'jquery';
import DateFormField from '../src';

const createDateField = (options = {}, values = {}) => {
  const opts = {
    jsxname: 'test-name',
    jsxlabel: 'test',
    jsxtype: 'single', // or cascade
    jsxfrom: '',
    jsxto: '',
    panel: 'day', // month or year or day
    autoMatchWidth: false,
  };
  return (
    <Form jsxvalues={values}>
      {React.createElement(DateFormField, { ...opts, ...options })}
    </Form>
  );
};

describe('DateFormField', () => {
  let div = null;

  beforeEach(() => {
    div = document.createElement('div');
    div.id = 'test-demo';
    document.body.appendChild(div);
  });

  afterEach(() => {
    document.body.removeChild(div);
    $('.uxcore').remove();
  });

  describe('Constructor Method', () => {
    it('call constructor correctly', () => {
      const w = mount(createDateField());
      expect(w.find(DateFormField).prop('jsxtype')).to.be('single');
      expect(w.find(DateFormField).prop('jsxfrom')).to.be('');
      expect(w.find(DateFormField).prop('jsxto')).to.be('');
      expect(w.find(DateFormField).prop('panel')).to.be('day');
      expect(w.find(DateFormField).prop('autoMatchWidth')).to.be(false);
    });
  });

  describe('Render Field Test', () => {
    it('render cascade date form field', (done) => {
      const w = mount(createDateField({ jsxtype: 'cascade' }));
      expect(w.find(DateFormField).prop('jsxtype')).to.be('cascade');
      ReactDOM.render(createDateField(), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-calendar-picker-input').length).to.be(1);
        done();
      }, 100);
    });

    it('render autoMatchWidth date form field', () => {
      const w = mount(createDateField({ jsxtype: 'cascade', autoMatchWidth: true }));
      const d = w.find(DateFormField);
      expect(d.prop('autoMatchWidth')).to.be(true);
      expect(d.node.resizeListenner.remove).to.be.a('function');
    });

    it('render showtime date form field', (done) => {
      const w = mount(createDateField({ jsxtype: 'cascade', showtime: true }));
      expect(w.find(DateFormField).prop('showtime')).to.be(true);
      ReactDOM.render(createDateField({ jsxtype: 'cascade', showtime: true }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-calendar-picker-input').length).to.be(2);
        done();
      }, 100);
    });

    it('render jsxlabel with date form field', (done) => {
      const w = mount(createDateField({ jsxlabel: 'date2333' }));
      expect(w.find(DateFormField).prop('jsxlabel')).to.be('date2333');
      ReactDOM.render(createDateField({ jsxlabel: 'date2333' }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-uxform-field .label-content').text()).to.be('date2333');
        done();
      }, 100);
    });

    it('render date form field with different formatter', () => {
      const w = mount(createDateField({ jsxtype: 'cascade', format: 'yyyy' }));
      expect(w.find(DateFormField).prop('format')).to.be('yyyy');
      w.unmount();
    });

    it('render the read view not editable', (done) => {
      ReactDOM.render(createDateField({
        jsxname: 'date',
        jsxlabel: 'date2333',
        jsxmode: Constants.MODE.VIEW,
      }, {
        date: '2012-11-12',
      }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-input').length).to.be(0);
        expect($('.kuma-uxform-field-core span').text())
          .to.be('2012-11-12');
        done();
      }, 100);
    });

    it('render the read view not editable with timer', (done) => {
      ReactDOM.render(createDateField({
        jsxlabel: 'date2333',
        jsxname: 'date',
        jsxmode: Constants.MODE.VIEW,
        showTime: true,
      }, {
        date: '2012-11-12',
      }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-input').length).to.be(0);
        expect($('.kuma-uxform-field-core span').text())
          .to.be('2012-11-12 08:00:00');
        done();
      }, 100);
    });

    it('render the date form field with initial value', (done) => {
      ReactDOM.render(createDateField({
        jsxlabel: 'date2333',
        jsxmode: Constants.MODE.VIEW,
      }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-input').length).to.be(0);
        done();
      }, 100);
    });

    it('render the date form field with initial value', (done) => {
      ReactDOM.render(createDateField({
        useFormat: true,
        jsxname: 'date',
        format: 'yyyy-MM-dd',
        jsxlabel: 'date2333',
        jsxmode: Constants.MODE.VIEW,
        jsxtype: 'cascade',
      }, {
        date: ['2012-11-12', '2012-12-12'],
      }), document.getElementById('test-demo'));
      setTimeout(() => {
        expect($('.kuma-input').length).to.be(0);
        expect($('.kuma-uxform-field-core span').text())
          .to.be('2012-11-12 - 2012-12-12');
        done();
      }, 100);
    });
  });

  describe('Event Test', () => {
    it('single date form field support onClick and onChange event to show picker panel', (done) => {
      ReactDOM.render(createDateField({ jsxlabel: 'date2333' }), document.getElementById('test-demo'));
      setTimeout(() => {
        $('.kuma-uxform-field input').click();
        setTimeout(() => {
          expect($('.kuma-calendar-picker').css('display')).to.be('block');
          done();
        }, 100);
      }, 100);
    });

    it('single date form field support onClick and onChange event to change value', (done) => {
      ReactDOM.render(createDateField({
        showTime: false,
        format: 'yyyy-MM-dd',
        jsxfrom: '',
        jsxto: '',
        jsxname: 'date',
        jsxlabel: '日期',
        locale: 'zh-cn',
      }), document.getElementById('test-demo'));
      setTimeout(() => {
        $('.kuma-uxform-field input').click();
        setTimeout(() => {
          const $date = $('.kuma-calendar-date').last();
          $date.click();
          setTimeout(() => {
            const $input = $('.kuma-calendar-picker-input .kuma-input');
            // set value is the latest date string
            expect($input.val()).to.be.a('string');
            done();
          }, 100);
        }, 100);
      }, 100);
    });

    it('cascade date form field support onClick and onChange event', (done) => {
      ReactDOM.render(createDateField({
        jsxlabel: 'date2333',
        jsxtype: 'cascade',
      }),
        document.getElementById('test-demo'));
      setTimeout(() => {
        $('.kuma-uxform-field input').click();
        setTimeout(() => {
          const $date = $('.kuma-calendar-date').last();
          $date.click();
          setTimeout(() => {
            const $input = $('.kuma-calendar-picker-input .kuma-input');
            // set value is the latest date string
            expect($input.val()).to.be.a('string');
            done();
          }, 100);
        }, 100);
      }, 100);
    });
  });
});


src/DateFormField.js
/**
 * DateFormField Component for uxcore
 * @author eternalsky
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */

const React = require('react');
const FormField = require('uxcore-form-field');
const Constants = require('uxcore-const');
const classnames = require('classnames');
const Calendar = require('uxcore-calendar');
const assign = require('object-assign');
const deepcopy = require('deepcopy');
const Formatter = require('uxcore-formatter');
const CalendarPanel = {
    month: Calendar.MonthCalendar,
    year: Calendar.YearCalendar,
    day: Calendar,
}

class DateFormField extends FormField {
    constructor(props) {
        super(props);
    }

    handleChange(value, format) {
        const me = this;
        const { useFormat } = me.props;
        let data;
        if (useFormat) {
            data = format;
        }
        else {
            data = value ? new Date(value).getTime() : null;
        }
        me.handleDataChange(data);
    }

    handleCascadeChange(i, value, format) {
        const me = this;
        let values = deepcopy(me.state.value) || [];
        const { useFormat } = me.props;
        let data;
        if (useFormat) {
           data = format; 
        }
        else {
            data = value ? new Date(value).getTime() : undefined;
        }
        values[i] = data;
        if (value) {
            if (i == 0 && !!values[1] && new Date(value).getTime() > new Date(values[1]).getTime()) {
                values.pop();
            }
            if (i == 1 && !!values[0] && new Date(value).getTime() < new Date(values[0]).getTime()) {
                values[0] = undefined;
            }
        }
        me.handleDataChange(values);
    }

    addSpecificClass() {
        let me = this;
        if (me.props.jsxprefixCls == "kuma-uxform-field") {
            let str = me.props.jsxprefixCls + " kuma-date-uxform-field" ;
            if (me.props.jsxtype == "cascade") {
                str += " kuma-cascade-date-uxform-field"
            }
            return str;
        }
        else {
            return me.props.jsxprefixCls
        }
    }

    processTime(time) {
        // if showTime is true or timePicker is set, we use time to compare
        // otherwise we use day to compare
        let me = this;
        let {showTime, timePicker} = me.props;
        if (showTime || timePicker) {
            return new Date(time).getTime();
        }
        else {
            return new Date(Formatter.date(time, 'YYYY-MM-DD')).getTime();
        }
    }

    renderField() {
        let me = this;
        let {onSelect, style, prefixCls, value, jsxtype, jsxfrom, jsxto, disabledDate, panel, ...others} = me.props;
        let from = !!jsxfrom ? me.processTime(jsxfrom) : -Infinity;
        let to = !!jsxto ? me.processTime(jsxto) : Infinity;
        let mode = me.props.jsxmode || me.props.mode;
        if (mode == Constants.MODE.EDIT) {
            const Panel = CalendarPanel[panel];
            if (jsxtype == "single") {
                return <Panel
                        value={me.state.value}
                        onSelect={me.handleChange.bind(me)}
                        disabledDate={disabledDate ? disabledDate : (current, value) => {
                            // if showTime is true or timePicker is set, we use time to compare
                            // otherwise we use day to compare
                            if (!current) {
                                return false;
                            }

                            return (me.processTime(current.getTime()) < from || me.processTime(current.getTime()) > to)
                        }}
                        {...others}/>
            }
            else if (jsxtype == "cascade") {
                let arr = [];
                let others1 = assign({}, others);
                let others2 = assign({}, others);
                if (me.state.value && me.state.value[0]) {
                    others1 = assign({}, others, {
                        value: me.state.value[0]
                    });
                }
                else {
                    others1 = assign({}, others, {
                        value: null
                    });
                }
                if (me.state.value && me.state.value[1]) {
                    others2 = assign({}, others, {
                        value: me.state.value[1]
                    });
                }
                else {
                    others2 = assign({}, others, {
                        value: null
                    });
                }
                arr.push(<Panel
                        key="calendar1"
                        onSelect={me.handleCascadeChange.bind(me, 0)}
                        disabledDate={(current, value) => {
                            if (!current) {
                                return false;
                            }
                            let now = me.processTime(current.getTime());
                            return (now < from || now > to)
                        }}
                        {...others1}/>);
                arr.push(<span key="split" className="kuma-uxform-split">-</span>)

                arr.push(<Panel
                        key="calendar2"
                        onSelect={me.handleCascadeChange.bind(me, 1)}
                        disabledDate={(current, value) => {
                            if (!current) {
                                return false;
                            }
                            let now = me.processTime(current.getTime());
                            let first = me.state.value ? me.state.value[0] : 0;
                            first = me.processTime(first);
                            return (now < from || now > to || now < first);
                        }}
                        {...others2}/>);
                return arr;
            }
        }
        else if (mode == Constants.MODE.VIEW) {
            let defautFormat = 'YYYY-MM-DD';
            if (me.props.showTime || me.props.timePicker) {
                defautFormat = 'YYYY-MM-DD HH:mm:ss';
            }
            if (jsxtype == "single") {
                return <span>{!!me.state.value ? Formatter.date(me.state.value, (me.props.format || defautFormat)) : ""}</span>
            }
            else {
                return <span>{!!me.state.value ? me.state.value.map((item) => {
                    return Formatter.date(item, (me.props.format || defautFormat))
                }).join(" - ") : ""}</span>
            }
        }
    }
}

DateFormField.displayName = "DateFormField";
DateFormField.propTypes = assign(FormField.propTypes, {
    jsxtype: React.PropTypes.string,
    panel: React.PropTypes.string,
    useFormat: React.PropTypes.bool,
});
DateFormField.defaultProps = assign(FormField.defaultProps, {
    locale: 'zh-cn',
    hasTrigger: true,
    jsxtype: 'single',
    panel: 'day',
    useFormat: false,
});
module.exports = DateFormField;


demo/DateFormFieldDemo.js
/**
 * DateFormField Component Demo for uxcore
 * @author eternalsky
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */

let classnames = require('classnames');

let DateFormField = require('../src');
let Form = require('uxcore-form/build/Form');
let Button = require('uxcore-button');
let Const = require('uxcore-const');

class Demo extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            mode: Const.MODE.EDIT
        }
    }

    handleModeChange() {
        let me = this;
        me.setState({
            mode: me.state.mode == Const.MODE.EDIT ? Const.MODE.VIEW : Const.MODE.EDIT
        })
    }

    handleValueGet() {
        let me = this;
        console.log(me.refs.form.getValues());
    }

    render() {
        let me = this;
        return (
            <div>
                <Form jsxmode={me.state.mode} ref="form">
                    <DateFormField showTime format="yyyy-MM-dd HH:mm:ss" jsxname="date" jsxlabel="日期" locale="zh-cn" />
                    <DateFormField jsxtype="cascade" useFormat format="yyyy-MM-dd" jsxname="format" jsxlabel="useFormat" locale="zh-cn" />
                    <DateFormField jsxtype="cascade" jsxname="casDate" jsxlabel="级联日期" format="yyyy/MM/dd" jsxto="2016-08-15" showDateInput />
                </Form>
                <Button onClick={me.handleModeChange.bind(me)}>切换模式</Button>
                <Button onClick={me.handleValueGet.bind(me)}>获取数据</Button>

            </div>
        );
    }
};

module.exports = Demo;
